app/src/main/java/com/example/sharewhatyoucanproject/donor/DonorActivity.kt

```diff
@@ -0,0 +1,155 @@
+import android.widget.ArrayAdapter
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import androidx.lifecycle.ViewModelProvider
+import com.example.sharewhatyoucanproject.databinding.ActivityDonorBinding
+import com.example.sharewhatyoucanproject.utils.PERMISSION_ALL
+import com.example.sharewhatyoucanproject.utils.foodTypes
+import com.example.sharewhatyoucanproject.utils.hasPermissions
+import com.example.sharewhatyoucanproject.utils.isGPSEnabled
+import com.example.sharewhatyoucanproject.utils.locationPermissions
+import com.example.sharewhatyoucanproject.utils.showToast
+import com.google.android.gms.location.LocationRequest
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.storage.FirebaseStorage
+import com.google.firebase.storage.StorageReference
+
+class DonorActivity : AppCompatActivity() {
+
+    lateinit var donorViewModel: DonorViewModel
+    private lateinit var binding: ActivityDonorBinding
+    lateinit var db: FirebaseFirestore
+    lateinit var storage: FirebaseStorage
+    lateinit var storageReference: StorageReference
+    lateinit var locationRequest: LocationRequest
+    lateinit var arrayAdapter: ArrayAdapter<String>
+    lateinit var circularProgressIndicator: com.google.android.material.progressindicator.CircularProgressIndicator
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityDonorBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        circularProgressIndicator = findViewById(R.id.progress_circular)
+        locationRequest = LocationRequest.create()
+        supportActionBar?.hide()
+        storage = FirebaseStorage.getInstance()
+        storageReference = storage.reference
+        db = FirebaseFirestore.getInstance()
+
+        donorViewModel = ViewModelProvider(
+            this,
+            DonorViewModelFactory(applicationContext, db, storageReference),
+        )[DonorViewModel::class.java]
+
+        donorViewModel.currentLocation.observe(this) {
+            it?.let {
+                donorViewModel.uploadImage()
+            }
+        }
+        donorViewModel.imageUrl.observe(this) {
+            it?.let { donorViewModel.saveData() }
+        }
+        donorViewModel.saveData.observe(this) {
+            it?.let {
+                if (it == "success") {
+                    applicationContext.showToast("Post Added")
+                    finish()
+                } else {
+                    circularProgressIndicator.visibility = View.GONE
+                    applicationContext.showToast(it)
+                }
+            }
+        }
+
+        binding.foodimg.setOnClickListener {
+            val intent = Intent()
+            intent.type = "image/*"
+            intent.action = Intent.ACTION_GET_CONTENT
+            startActivityForResult(Intent.createChooser(intent, "Select Picture"), 0)
```

Comment:
we usually make a const value for the request code so that it's obvious what it is. You can do that by making a `companion object` in this activity:
```
companion object {
  const val REQUEST_CODE_IMAGE = 0
}
```